### PR TITLE
convert cgsched.c to D

### PR DESCRIPTION
**Pull request by WalterBright**

### CODE REVIEW

File: src/dmd/backend/cgsched.d [hunk 1]
```diff
@@ -2122,147 +2227,136 @@ if (c2->IEVpointer1 + sz2 <= c1->IEVpointer1) printf("t5\n");
...
             }
             {   code *cf;
-                cf = gen2(NULL,0xD9,0xC8 + ci->fxch_post);
+                cf = gen2(null,0xD9,0xC8 + ci.fxch_post);
                 *pc = cf;
-                pc = &code_next(cf);
+                pc = &cf.next;
             }
         }
     L1: ;
```
Comment (wilzbach) [inline src/dmd/backend/cgsched.d]: The newly added [DScanner](https://github.com/dlang-community/D-Scanner) (static source code analyzer and linter) complains about this empty declaration here.

File: src/dmd/backend/cgsched.d [hunk 2]
```diff
@@ -3006,70 +3106,73 @@ code *peephole(code *cstart,regm_t scratch)
...
+            default:
+                break;
         }
         continue;
 Lnop:
-        c1->Iop = NOP;
+        c1.Iop = NOP;
         c1 = cnext(c1);
         goto Ln;
     }
```
Comment (wilzbach) [inline src/dmd/backend/cgsched.d]: And below there's another instance of an empty declaration.

File: src/dmd/backend/cgsched.d [hunk 3]
```diff
@@ -188,485 +234,497 @@ enum
...
         // Could generate an exception we want to catch
-        mAX|EA|N,       F|mAX|N,        // IDIV
-#else
-        mAX|EA, F|mAX,          // IDIV
-#endif
+        //mAX|EA|N,       F|mAX|N,        // IDIV
 
-};
+      [ mAX|EA, F|mAX ],          // IDIV
+    ]
```
Comment (wilzbach) [inline src/dmd/backend/cgsched.d]: It only uses four top-level elements, not 8.

**Comment by ZombineDev** [comment]
The BuildKite CI fails at the style check step with the following error:
```
Running DScanner
--
  | ../generated/linux/release/64/dscanner-3a859d39c4b59822b1bc0452b3ddcd598ef390a2/dsc --config .dscanner.ini --styleCheck dmd -I.
  | dmd/backend/cgsched.d(2343:9)[warn]: Empty declaration
  | dmd/backend/cgsched.d(3179:5)[warn]: Empty declaration
  | posix.mak:617: recipe for target 'dscanner' failed
  | make[1]: *** [dscanner] Error 1
  | make[1]: Leaving directory '/var/lib/buildkite-agent/builds/dime-2/dlang/dmd/src'
  | posix.mak:53: recipe for target 'style' failed
  | make: *** [style] Error 2
```

Removing the semicolons here:
https://github.com/dlang/dmd/blob/9a7bde5a12d327ce21b287897d389297ac66cbb4/src/dmd/backend/cgsched.d#L2342-L2344
and here:
https://github.com/dlang/dmd/blob/9a7bde5a12d327ce21b287897d389297ac66cbb4/src/dmd/backend/cgsched.d#L3178-L3180
should fix the problem.

**Review by wilzbach — APPROVED** [review]

**Comment by wilzbach** [comment]
> Removing the semicolons here:

Quickly amended this to the PR.

**Review by PetarKirov — APPROVED** [review]

File: src/dmd/backend/cgsched.d [hunk 4]
```diff
@@ -2122,147 +2227,136 @@ if (c2->IEVpointer1 + sz2 <= c1->IEVpointer1) printf("t5\n");
...
-/******************************
- */
-
-code **Schedule::assemble(code **pc)
+code **assemble(code **pc)  // reassemble scheduled instructions
 {
     code *c;
 
-#ifdef DEBUG
+    debug
```
Comment (JinShil) [inline src/dmd/backend/cgsched.d]: I would prefer a pair of curly braces to accompany this.

**Review by JinShil — APPROVED** [review]
Translation looks good to me.

File: src/dmd/backend/cgsched.d [hunk 5]
```diff
@@ -2122,147 +2227,136 @@ if (c2->IEVpointer1 + sz2 <= c1->IEVpointer1) printf("t5\n");
...
+            cf = gen2(null,0xD9,0xC8 + ci.fxch_pre);
             *pc = cf;
-            pc = &code_next(cf);
+            pc = &cf.next;
         }
 
         *pc = c;
         do
         {
             assert(*pc != code_next(*pc));
```
[the remark this answers is not in the document]
Comment (WalterBright) [reply src/dmd/backend/cgsched.d]: It's used as an rvalue there, not an lvalue.

File: src/dmd/backend/cgsched.d (hunk 3 above)
Comment (WalterBright) [reply src/dmd/backend/cgsched.d]: So does the original!

**Pull request merged**